src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
+			return $result;
+		}
+
+		// Already processed.
+		if ( in_array( $hook_extra['plugin'], self::$processed, true ) ) {
+			return $result;
+		}
+
+		error_log( $hook_extra['plugin'] . ' processing...' );
+
+		/**
+		 * This possibly helps to avoid a potential race condition on servers that may start to
+		 * process the next plugin for auto-updating before the handler can pick up an error from
+		 * the previously processed plugin.
+		 */
+		sleep( 2 );
+
+		$this->no_error     = false;
+		static::$current    = get_site_transient( 'update_plugins' );
+		$this->handler_args = array(
+			'handler_error' => '',
+			'result'        => $result,
+			'hook_extra'    => $hook_extra,
+		);
+
+		// Register exception and shutdown handlers.
+		$this->initialize_handlers();
+
+		self::$processed[] = $hook_extra['plugin'];
+
+		if ( is_plugin_inactive( $hook_extra['plugin'] ) ) {
+			// Working parts of plugin_sandbox_scrape().
+			wp_register_plugin_realpath( WP_PLUGIN_DIR . '/' . $hook_extra['plugin'] );
+			include WP_PLUGIN_DIR . '/' . $hook_extra['plugin'];
+		}
+
+		// Needs to run for both active and inactive plugins. Don't ask why, just accept it.
+		$this->check_plugin_for_errors( $hook_extra['plugin'] );
+		error_log( $hook_extra['plugin'] . ' auto updated.' );
+
+		return $result;
+	}
+
+	/**
+	 * Checks a new plugin version for errors.
+	 *
+	 * If an error is found, the previously installed version will be reinstalled
+	 * and an email will be sent to the site administrator.
+	 *
+	 * @param string $plugin The plugin to check.
+	 *
+	 * @throws \Exception If errors are present.
+	 *
+	 * @return void
+	 */
+	private function check_plugin_for_errors( $plugin ) {
+		global $wp_filesystem;
+
+		if ( file_exists( ABSPATH . '.maintenance' ) ) {
+			$wp_filesystem->delete( ABSPATH . '.maintenance' );
+		}
+
+		$nonce    = wp_create_nonce( 'plugin-activation-error_' . $plugin );
+		$response = wp_remote_get(
+			add_query_arg(
+				array(
+					'action'   => 'error_scrape',
+					'plugin'   => $plugin,
+					'_wpnonce' => $nonce,
+				),
+				admin_url( 'plugins.php' )
+			),
+			array( 'timeout' => 60 )
+		);
+
+		if ( is_wp_error( $response ) ) {
+			// If it isn't possible to run the check, assume an error.
+			throw new \Exception( $response->get_error_message() );
+		}
+
+		$code           = wp_remote_retrieve_response_code( $response );
+		$body           = wp_remote_retrieve_body( $response );
+		$this->no_error = 200 === $code;
+
+		if ( str_contains( $body, 'wp-die-message' ) || 200 !== $code ) {
+			$error = new WP_Error(
+				'new_version_error',
+				sprintf(
+					/* translators: %s: The name of the plugin. */
+					__( 'The new version of %s contains an error' ),
+					get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
+				)
+			);
+			throw new \Exception( $error->get_error_message() );
```

Comment:
Should this just send the message directly to `new Exception()` instead?

Reply:
something tells me `throw new Exception( new WP_Error()...` might be a little less clear.

Reply:
I meant more:

```php
throw new Exception(
	sprintf(
		/* translators: %s: The name of the plugin. */
		__( 'The new version of %s contains an error' ),
		get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
	)
);
```